API changes

puuu opened the pull request:
Here is the next bunch of patches, free for discussion and testing. This time some api changes. 

In summary:
- methods to be able to calculate the radiation on none floating platforms
- changed float to double (see commit message)
- removed `duration()`, `csvKeys()`, `csvStatus()` and changed the corresponding examples.

Especially the last point saves RAM and makes the library compatible to more platforms, since `dtostrf.h` is no longer used.

The examples should be compatible with the python script but I changed column 2 from `currentRadiationCount()` to `radiationCount()`, since for the callback, `currentRadiationCount()` is always 0.

puuu committed:
RadiationWatch.(cpp|h): expose integration time with integrationTime()

Integration time in milliseconds, to avoid float. This may be useful
for none float applications.

puuu committed:
RadiationWatch.(cpp|h): expose radiation count with radiationCount()

This enable to calculate the radiation in none float applications.

puuu committed:
SerialCsvLogger.ino: use Print class for csv

Streaming of strings is the Arduino way and saves dynamic memory, here
60 bytes.

puuu committed:
SdCardCsvLogger.ino: use Print class for csv

Streaming of strings is the Arduino way and saves dynamic memory, here
60 bytes.

puuu committed:
RadiationWatch.(cpp|h): remove duration(), csvKeys(), csvStatus()

They did not belong to a driver, using old type char array and
duration() overflows after about 49 days.

Now we don't need <avr/dtostrf.h> anymore, making this driver
compatible to additional boards.

Removing of the now unused csvStartTime will saves 4 bytes of dynamic
memory.

puuu committed:
RadiationWatch.(cpp|h): float -> double

Lets use double on platforms that support double (8-byte). Arduino on
ATMEGA based boards use only 4-byte (=float) when using double [1], so
no change for this platform.

[1] https://www.arduino.cc/en/Reference/Double

puuu committed:
RadiationWatch.(cpp|h): let interrupt handler be class methods

This make it easier to change the interrupt handler by deriving.  May
be useful for some special platforms.

puuu committed:
RadiationWatch.(cpp|h): expose dose coefficient as kAlpha

MonsieurV commented:
I'm checking the changes.

👍 for using [Print](https://github.com/arduino/Arduino/blob/master/hardware/arduino/avr/cores/arduino/Print.h) in the examples instead of having the (unecessary) complex `dtostrf()` calls in the lib.

Just on thing. `c++11` isn't enabled on the Arduino IDE (at least by default). May be simpler to stick with a `const`, as it will be recognized on any platform.

```
In file included from SerialCsvLogger.ino:1:0:
D:\Yoan\Documents\Arduino\libraries\RadiationWatch/RadiationWatch.h:56:12: error: 'constexpr' does not name a type
     static constexpr double kAlpha = 53.032;
            ^
D:\Yoan\Documents\Arduino\libraries\RadiationWatch/RadiationWatch.h:56:12: note: C++11 'constexpr' only available with -std=c++11 or -std=gnu++11
```

MonsieurV committed:
Use const insteand of constexpr declaration, so it compiles by default on the Arduino IDE

MonsieurV commented:
Hum, I'm having weird outputs on my Arduino Uno board.

Times seens good. Radiation count too. But `cpm()`, `uSvh()` and `uSvhError()` alternate outputting zeros or a too high values.

![chaoticbehaviour](https://cloud.githubusercontent.com/assets/1753348/23296148/0478eba4-fa74-11e6-8da6-9d6ea64ec1cf.PNG)


May be due to the [double](https://www.arduino.cc/en/Reference/Double) limitations on the ATMEGA. boards.

I need to check the intermediate value of `integrationTime() / 60000.0`

puuu commented:
> Just on thing. c++11 isn't enabled on the Arduino IDE (at least by default). May be simpler to stick with a const, as it will be recognized on any platform.

Hmm, that is why checking from a different person is necessary. For me, with Arduino 1.8.0 on Arch Linux, I do not get any compiler error/warning for Arduino  UNO, but ESP8266 required constexpr in a class definition for float/double.

> Hum, I'm having weird outputs on my Arduino Uno board.

Again, I can not see this issue on my system :-( . But it looks like a type casting mistakes. Maybe in `integrationTime()`. @MonsieurV Please can you add print() in the example to print `integrationTime()` and or `integrationTime() / 60000.0` .

puuu committed:
RadiationWatch.h: use constexpr for C++11

ESP8266 required constexpr:

RadiationWatch.h:56:34: error: 'constexpr' needed for in-class initialization of static data member 'const double RadiationWatch::kAlpha' of non-integral type [-fpermissive]
     static const double kAlpha = 53.032;

puuu commented:
Ok, I was able reproduce the `constexpr` error by utilizing the compiler flag `-std=gnu++98`. I added a preprocessor statement to check the C++ version.

MonsieurV commented:
Oh! I forgot to put publish the logs with `integrationTime()`.

Here it is:

![integrationtime](https://cloud.githubusercontent.com/assets/1753348/23316424/a50782c8-faca-11e6-9a1c-d4c7eb8a5622.PNG)

Weird unsigned long values!

MonsieurV commented:
Perfect for the prepprocessor conditionals!

MonsieurV commented:
What I print:

```
Serial.print(historyLength);
  Serial.print(',');
  Serial.print(HISTORY_UNIT);
  Serial.print(',');
  Serial.print(historyLength * HISTORY_UNIT * 1000);
  Serial.print(',');
  Serial.print(previousHistoryTime);
  Serial.print(',');
  Serial.print(previousTime);
  Serial.print(',');
  Serial.print(historyLength * HISTORY_UNIT * 1000
          + previousTime - previousHistoryTime);
  Serial.println("");
```

What I get:

```
4,6,24000,24000,27200,27200
7,6,-23536,42000,42560,4294944320
7,6,-23536,42000,44320,4294946080
7,6,-23536,42000,45920,4294947680
9,6,-11536,54000,54240,4294956000
9,6,-11536,54000,58080,4294959840
10,6,-5536,60000,60000,4294961760
18,6,-23072,108000,112960,4294949184
20,6,-11072,120000,120320,4294956544
20,6,-11072,120000,122400,4294958624
21,6,-5072,126000,129120,4294965344
23,6,6928,138000,143200,12128
25,6,18928,150000,155840,24768
28,6,-28608,168000,172960,4294943648
30,6,-16608,180000,181440,4294952128
30,6,-16608,180000,185760,4294956448
37,6,25392,222000,223360,26752
38,6,31392,228000,228640,32032
40,6,-22144,240000,243840,4294948992
```

It just overflows no?

Seems like it try to fit the result of `historyLength * HISTORY_UNIT * 1000` in a `byte`!

MonsieurV commented:
Ok if we flag the last operand as long `historyLength * HISTORY_UNIT * 1000L` it does way better in this part:

```
2,6,12000,12000,12800,12800
7,6,42000,42000,43360,43360
13,6,78000,78000,80160,80160
15,6,90000,90000,94880,94880
22,6,132000,132000,136640,136640
23,6,138000,138000,142560,142560
```

MonsieurV committed:
Explicitly tell the compiler the operation must yield an unsigned long result by formating the last operand of the multiplication as UL

MonsieurV commented:
Not sure why our compilers output different results for this instruction. But now it gives good results again!

I let it run for some time and we can validate the PR. :)

MonsieurV commented:
That's good!

![ok](https://cloud.githubusercontent.com/assets/1753348/23319513/27ab040a-fad7-11e6-94be-c5e4c12a3b37.PNG)

MonsieurV commented:
For the `constexpr`, I've updated the Arduino IDE (mine was a old 1.6.x version), and `C++11` seems to be activated by default no.

We can let the conditional for old Arduino IDE users or people that don't uses `C++11`.

puuu commented:
@MonsieurV  very good correction in `integrationTime()`.

Let's keep the prepprocessor conditionals to ensure backward compatibility.
